tests/TestNoseHooverThermostat.h

```diff
@@ -0,0 +1,293 @@
+                minDist = std::min(minDist, sqrt(delta.dot(delta)));
+                delta = pos2-positions[j];
+                minDist = std::min(minDist, sqrt(delta.dot(delta)));
+            }
+            if (minDist > 0.15) {
+                positions[2*i+0] = pos1;
+                positions[2*i+1] = pos2;
+                break;
+            }
+        }
+    }
+    bool simpleConstruct = true;
+    double temperature = 300; // kelvin
+    double collisionFrequency = 25; // 1/ps
+    int numMTS = 3;
+    int numYS = 3;
+    int chainLength = 5;
+    auto integrator = simpleConstruct ? NoseHooverIntegrator(temperature, collisionFrequency, 0.001, chainLength, numMTS, numYS)
+                                      : NoseHooverIntegrator(0.001);
+    if (!simpleConstruct)
+        integrator.addThermostat(temperature, collisionFrequency, chainLength, numMTS, numYS);
+    Context context(system, integrator, platform);
+    context.setPositions(positions);
+    context.setVelocitiesToTemperature(temperature);
+    integrator.step(1500);
+
+    int nSteps = 2000;
+    double mean_temp = 0.0;
+    std::vector<double> energies(nSteps);
+    for (int i = 0; i < nSteps; ++i) {
+        integrator.step(1);
+        State state = context.getState(State::Energy | (constrain ? State::Positions : 0));
+        if (constrain) {
+            auto positions = state.getPositions();
+            for(int i = 0; i < numMolecules; ++i) {
+                Vec3 delta = positions[2*i+1] - positions[2*i];
+                double dR2 = delta.dot(delta);
+                ASSERT_EQUAL_TOL(bondLengthSquared, dR2, 1e-4);
+            }
+        }
+        double KE = state.getKineticEnergy();
+        double PE = state.getPotentialEnergy();
+        double time = state.getTime();
+        double instantaneous_temperature = 2 * KE / (BOLTZ * numDOF);
+        mean_temp = (i*mean_temp + instantaneous_temperature)/(i+1);
+        double energy = KE + PE + integrator.computeHeatBathEnergy();
+#if DEBUG
+        if(i % 10 == 0) std::cout << " Time: " << std::setw(4) << time 
+                                  << " Temperature: " << std::setw(8) << std::setprecision(3) << temperature
+                                  << " Mean Temperature: " << std::setw(8) << std::setprecision(3) << mean_temp
+                                  << " Total Energy: " << std::setw(12) << std::setprecision(8) << energy << std::endl;
+#endif
+        energies[i] = energy;
+    }
+    double sum = std::accumulate(energies.begin(), energies.end(), 0.0);
+    double mean = sum / energies.size();
+    double sq_sum = std::inner_product(energies.begin(), energies.end(), energies.begin(), 0.0);
+    double std = std::sqrt(sq_sum / energies.size() - mean * mean);
+    double relative_std = std / mean;
+    //std::cout << "Relative STD of energy " << relative_std << std::endl;
+    // Check mean temperature
+    ASSERT_EQUAL_TOL(temperature, temperature, 1e-2);
```

Comment:
I think one of those should be `mean_temp`?

Reply:
ooops